### ISSUE TITLE
Use ctrl-v to visually mark some block, and then `:` to execute a command over that block

### ISSUE THREAD

**Issue by beyarkay** [neovim/neovim#22297]
### Problem

(Originally asked about solutions [here](https://stackoverflow.com/questions/75470770/neovim-substitute-over-block-regular-visual-range-and-not-line-visual-range/75484225#75484225))

In vim/neovim, ranges are only line-wise. I'd like to be able to perform commands on column- and line-wise ranges. There are more examples in the Stack overflow post, but I find it odd that [the vim help pages](https://vimhelp.org/visual.txt.html#%3Avisual_example) talk about line-based ranges as though column-based ranges were planned in the future:

```
							*:visual_example*
Currently the ":" command works on whole lines only.  When you select part of
a line, doing something like ":!date" will replace the whole line.  If you
want only part of the line to be replaced you will have to make a mapping for
it.  In a future release ":" may work on partial lines.
```

Is there any tracking issue for the comment "In a future release ":" may work on partial lines"? If not, I felt there should be one. It would be a big help.

There is a plugin that partially solves this: [vis](https://github.com/vim-scripts/vis), although it's not as ergonomic as it could be.

### Expected behavior

If a block of text were selected using `ctrl-v` (or even just a part of a line of text), then pressing `:` should auto-fill the partial-line range selector: ```:`<,`>``` instead of it's current behaviour, which auto-fills the line-based range selector: `:'<,'>`. I'm not sure if it's feasible (or possible) to migrate all line-based commands to the new selector.

**Comment by justinmk** [comment]
> auto-fill the partial-line range selector: `` :`<,`> ``

This is in vim's todo.txt I believe. It is relevant for all commands, not just `:` ...

**Comment by justinmk** [comment]
otoh, it's pretty easy to make a plugin that : 

1. copies the selected block, 
2. pastes it into a new scratch buffer
3. sets 'filetype' to match the original buffer
4. executes the command

Not perfect but could be pretty useful.

**Comment by ad-chaos** [comment]
You can use the `\%V` atom to only substitute in visual mode (thanks to [chribra](https://github.com/vim/vim/issues/4272#issuecomment-483967346))

**Comment by justinmk** [comment]
> use the `\%V` atom

That is a special case for `:substitute` but doesn't answer the general idea of being able to pass a charwise or blockwise text to commands.

**Comment by beyarkay** [comment]
Good catch about blockwise selection being different from charwise selection. Maybe something like ``` :`a,`bfoo ``` for charwise and ```:1.3,$.5foo``` for blockwise?

I'm not fully up to date on nvims goals, does something like this belong in the main program or is it more suitable for a plugin? I don't have experience with plugins/nvim internals but would be keen to get my hands dirty

**Comment by justinmk** [comment]
This would probably need to be a core change

**Comment by beyarkay** [comment]
Okay, who would I speak to in order to get an idea about what the definition of done is? Maybe this would need to be broken into a charwise and a blockwise change, although I'm not sure what your status quo is.

**Comment by justinmk** [comment]
> what the definition of done is? Maybe this would need to be broken into a charwise and a blockwise change

Yeah, let's start with charwise first. I'd say "done" for charwise is:

# Proposal

- Support backtick ranges pass column info to the command
  ```
  :`<,`>foo
  :`a,`bfoo
  ```
    - Note: it's vaguely possible that parsing a backtick range might conflict with `:help backtick-expansion`. Probably isn't an issue because `:help backtick-expansion` does not happen in the "range" part (the start) of a command.
- Column info is available as `<col1>` , `<col2>` (like `:help <line1>`) 
    - For user-defined command, `:command -addr=lines`
- Support "x.y" ranges pass column info to the command
  ```
  :1.2,2.2 foo
  ```
- Update `nvim_parse_cmd()` , `nvim_cmd()`
- Add charwise support to a few common Ex commands:
    - `:read`
    - `:write`
    - `:substitute`
    - `:global`
    - ...
    - Document the rest.
- Update docs: `:help :range`,

**Comment by gpanders** [comment]
> * Support "x.y" ranges pass column info to the command
>       ```
>       :1.2,2.2 foo
>       ```

Suggestion: if line number is omitted, it should default to cursor line (which is already a plain `.`). So e.g. `:.2,.4foo` is "columns 2-4 of current cursor line".

**Comment by 616b2f** [comment]
> > * Support "x.y" ranges pass column info to the command
> >   `:1.2,2.2 foo`
> 
> Suggestion: if line number is omitted, it should default to cursor line (which is already a plain `.`). So e.g. `:.2,.4foo` is "columns 2-4 of current cursor line".

I am currently looking into this, it seems that `:.2,.4foo` is equivalent to `:.+2,.+4foo`  which has already a meaning, here the relevant parts:

https://github.com/neovim/neovim/blob/f05a6666cfcb2cd95d860620748cb0fa58b23aba/src/nvim/ex_docmd.c#L3592
https://github.com/neovim/neovim/blob/f05a6666cfcb2cd95d860620748cb0fa58b23aba/src/nvim/ex_docmd.c#L3625

**Comment by justinmk** [comment]
That was a nice-to-have, not really part of the main goal here?

**Comment by 616b2f** [comment]
> But this could be for all commands, not just `:s`. And I agree that backtick mark notation should work charwise:
> 
> ```
> :`<,`>foo
> :`a,`bfoo
> ```
> 
> Note that charwise is different than blockwise. Would need some way to disambiguate that.

@justinmk could you explain or give me a hint where to find the info, what is the difference between blockwise and charwise that you mention here?

**Comment by SebasF1349** [comment]
It's the difference between `v` (charwise - contiguous chars) and `ctrl-v` (blockwise - columns of chars), `:h visual-start`.
